graph.go

```diff
@@ -718,3 +720,238 @@ func (g *Graph) validateTransactionParents(tx *Transaction) error {
+
+	queue := AcquireQueue()
+	defer ReleaseQueue(queue)
+
+	for _, parentID := range tx.ParentIDs {
+		if parent := g.FindTransaction(parentID); parent != nil {
+			queue.PushBack(parent)
+		}
+
+		visited[parentID] = struct{}{}
+	}
+
+	hasher, _ := blake2b.New256(nil)
+
+	var depthCounter uint64
+	var lastDepth = tx.Depth
+
+	for queue.Len() > 0 {
+		popped := queue.PopFront().(*Transaction)
+
+		if popped.Depth != lastDepth {
+			lastDepth = popped.Depth
+			depthCounter++
+		}
+
+		// If we exceed the max eligible depth we search for candidate
+		// validators to reward from, stop traversing.
+
+		if depthCounter >= sys.MaxDepthDiff {
+			break
+		}
+
+		// Filter for all ancestral transactions not from the same sender,
+		// and within the desired graph depth.
+
+		if popped.Sender != tx.Sender {
+			stake, _ := ReadAccountStake(snapshot, popped.Sender)
+
+			if stake > sys.MinimumStake {
+				candidates = append(candidates, popped)
+				stakes = append(stakes, stake)
+
+				totalStake += stake
+
+				// Record entropy source.
+				if _, err := hasher.Write(popped.ID[:]); err != nil {
+					return errors.Wrap(err, "stake: failed to hash transaction ID for entropy source")
+				}
+			}
+		}
+
+		for _, parentID := range popped.ParentIDs {
+			if _, seen := visited[parentID]; !seen {
+				if parent := g.FindTransaction(parentID); parent != nil {
+					queue.PushBack(parent)
+				}
+
+				visited[parentID] = struct{}{}
+			}
+		}
+	}
+
+	// If there are no eligible rewardee candidates, do not reward anyone.
+
+	if len(candidates) == 0 || len(stakes) == 0 || totalStake == 0 {
+		return nil
+	}
+
+	entropy := hasher.Sum(nil)
+	acc, threshold := float64(0), float64(binary.LittleEndian.Uint64(entropy)%uint64(0xffff))/float64(0xffff)
+
+	var rewardee *Transaction
+
+	// Model a weighted uniform distribution by a random variable X, and select
+	// whichever validator has a weight X ≥ X' as a reward recipient.
+
+	for i, tx := range candidates {
+		acc += float64(stakes[i]) / float64(totalStake)
+
+		if acc >= threshold {
+			rewardee = tx
+			break
+		}
+	}
+
+	// If there is no selected transaction that deserves a reward, give the
+	// reward to the last reward candidate.
+
+	if rewardee == nil {
+		rewardee = candidates[len(candidates)-1]
+	}
+
+	rewardeeBalance, _ := ReadAccountReward(snapshot, rewardee.Sender)
+	WriteAccountReward(snapshot, rewardee.Sender, rewardeeBalance+fee)
+
+	if logging {
+		logger := log.Stake("reward_validator")
+		logger.Info().
+			Hex("creator", tx.Creator[:]).
+			Hex("recipient", rewardee.Sender[:]).
+			Hex("creator_tx_id", tx.ID[:]).
+			Hex("rewardee_tx_id", rewardee.ID[:]).
+			Hex("entropy", entropy).
+			Float64("acc", acc).
+			Float64("threshold", threshold).Msg("Rewarded validator.")
+	}
+
+	return nil
+}
+
+func (g *Graph) CollapseTransactions(accounts LikeAccounts, round uint64, latestRound *Round, root *Transaction, end *Transaction, logging bool) (*CollapseResults, error) {
```

Comment:
Again, would be good to hide/not expose it s.t. `collapseTransactions`. Also, why move it to `graph.go`?

Reply:
Same as above.